docs/SPEC-composition.md

```diff
@@ -0,0 +1,173 @@
+Flux Composition
+==
+
+The composition capabilities of the flux LSP server enable clients to create a query builder experience without deep knowledge of the flux internals. Composition will create a "Composition-owned" statement in a given flux file, and will be able to add/remove filter calls to that statement. The composition rules apply as follows:
+
+  - Composition must never generate a query that cannot be executed.
+  - Composition should never generate a query that a user with sufficient flux knowledge would never write, e.g. `filter(fn: (r) => r.foo == "bar") |> filter(fn: (r) => r.foo == "bar")`
+  - Composition should never generate a query that results in immediate diagnostic messages.
+  - In the event of conflicts or logical ambiguity, Composition should default to erroring rather than introducing complex assumption logic to attempt recovery.
+
+Given these rules, this document will not attempt to document the resulting flux code generated from each of these requests, as the resulting flux could change as requirements evolve and the corresponding flux would become outdated.
+
+Composition Initialize
+--
+
+The composition initialize request is sent from the client to the server to identify a file that will be used to compose queries. This call must be made before any other Composition requests, otherwise those requests will error. It will result in a `workspace/applyEdit` request from the server to the client with the associated changes to the file. The request can also be used to reset the composition query state by calling it again, and can be considered idempotent.
```

Comment:
The LSP protocol is send/response. But is it capable of allowing 2 responses per a single message?

Asking in case we brainstorm ideas for having the LSP send additional information to prompt the monaco-editor.

Reply:
`workspace/applyEdit` is a server-to-client send/response, like diagnostics work.

Reply:
yes. That's a 1-to-1.

I'm thinking about next steps...when we want to also kick the monaco-editor to do additional things, after the `workspace/applyEdit`.

Reply:
scratch that. Server initiated events. Need more coffee. snort.